File: src/Hooks/useEvent.lua
local CONNECT_METHODS = { "Connect", "connect", "on" }
local DISCONNECT_METHODS = { "Destroy", "destroy", "Disconnect", "disconnect" }

local function connect(obj, callback)
        if typeof(obj) == "RBXScriptSignal" then
                return obj:Connect(callback)
        elseif type(obj) == "table" then
                for _, method in CONNECT_METHODS do
			if type(obj[method]) ~= "function" then
				continue
			end

			return obj[method](obj, callback)
		end
        end

        error("Invalid signal object.", 3)
end

local function createDisconnect(connection)
        return function()
                if connection == nil then
                        return
                end

                for _, method in DISCONNECT_METHODS do
                        if type(connection[method]) ~= "function" then
                                continue
                        end

                        connection[method](connection)
                        return
                end

                error("Invalid connection object.", 4)
        end
end

--[=[
        Memoizes and connects a signal object to a callback listener.

        ```lua
        local function HookedComponent(props, hooks)
                local counter, setCount = useState(0)

                useEvent(mySignal, function(n)
                        -- Assuming that the signal has a number parameter,
                        -- the counter is incremented by `n` every time the signal fires.
                        setCount(function(current)
                                return current + n
                        end)
                end, { counter })(hooks)

                return Roact.createElement("TextLabel", {
                        -- ...
                        Text = "Count: " .. counter,
                })
        end
        ```

        @function useEvent
        @within Hooks
        @param signal RBXScriptSignal
        @param onEvent ((...any) -> ())
        @param dependencies {any} | nil
        @return HookCreator<R>
]=]
local function useEvent(signal, onEvent, deps)
        return function(hooks)
                hooks.useEffect(function()
                        return createDisconnect(connect(signal, onEvent))
                end, deps)
        end
end

return useEvent

File: typings/useEvent.ts
import { HookCreator } from ".";

interface SignalWithConnect {
        Connect(callback: Callback): { Disconnect(): void } | { Destroy(): void };
}

interface SignalWithCamelConnect {
        connect(callback: Callback): { disconnect(): void } | { destroy(): void };
}

interface SignalWithOn {
        on(callback: Callback): { disconnect(): void } | { destroy(): void };
}

type SignalLike = SignalWithConnect | SignalWithCamelConnect | SignalWithOn;

type InferSignalCallback<S extends SignalLike> = Parameters<
        S extends SignalWithConnect
                ? S["Connect"]
                : S extends SignalWithCamelConnect
                ? S["connect"]
                : S extends SignalWithOn
                ? S["on"]
                : never
>[0];

/**
 * Memoizes and connects a signal object to a callback listener.
 * @param signal The signal object.
 * @param onEvent The callback to execute when the event is fired.
 */
declare function useEvent<E extends SignalLike>(
        signal: E,
        onEvent: InferSignalCallback<E>,
): HookCreator<void>;

/**
 * Memoizes and connects a signal object to a callback listener.
 * @param signal The signal object.
 * @param onEvent The callback to execute when the event is fired.
 * @param dependencies The dependencies.
 */
declare function useEvent<E extends SignalLike>(
        signal: E,
        onEvent: InferSignalCallback<E>,
        dependencies?: Array<any>
): HookCreator<void>;

export = useEvent;